service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java

```diff
@@ -18,19 +18,112 @@
+ * contextual table-level properties.
+ *
+ * <p>This class acts as a translation layer between Polaris properties and the properties required
+ * by Iceberg's {@link FileIO}. For example, it evaluates storage actions and retrieves subscoped
+ * credentials to initialize a {@link FileIO} instance with the most limited permissions necessary.
+ */
 @ApplicationScoped
 @Identifier("default")
 public class DefaultFileIOFactory implements FileIOFactory {
+
+  private final RealmEntityManagerFactory realmEntityManagerFactory;
+  private final MetaStoreManagerFactory metaStoreManagerFactory;
+  private final PolarisConfigurationStore configurationStore;
+
+  @Inject
+  public DefaultFileIOFactory(
+      RealmEntityManagerFactory realmEntityManagerFactory,
+      MetaStoreManagerFactory metaStoreManagerFactory,
+      PolarisConfigurationStore configurationStore) {
+    this.realmEntityManagerFactory = realmEntityManagerFactory;
+    this.metaStoreManagerFactory = metaStoreManagerFactory;
+    this.configurationStore = configurationStore;
+  }
+
+  @Override
+  public FileIO loadFileIO(
+      @Nonnull RealmId realmId,
+      @Nonnull String ioImplClassName,
+      @Nonnull Map<String, String> properties) {
+    return loadFileIOInternal(ioImplClassName, properties);
+  }
+
   @Override
-  public FileIO loadFileIO(String impl, Map<String, String> properties) {
-    return CatalogUtil.loadFileIO(impl, properties, new Configuration());
+  public FileIO loadFileIO(
+      @Nonnull RealmId realmId,
+      @Nonnull String ioImplClassName,
+      @Nonnull Map<String, String> properties,
+      @Nonnull TableIdentifier identifier,
+      @Nonnull Set<String> tableLocations,
+      @Nonnull Set<PolarisStorageActions> storageActions,
+      @Nonnull PolarisResolvedPathWrapper resolvedEntityPath) {
+    PolarisEntityManager entityManager =
+        realmEntityManagerFactory.getOrCreateEntityManager(realmId);
+    PolarisCredentialVendor credentialVendor =
+        metaStoreManagerFactory.getOrCreateMetaStoreManager(realmId);
+    PolarisMetaStoreSession metaStoreSession =
+        metaStoreManagerFactory.getOrCreateSessionSupplier(realmId).get();
+
+    // Get subcoped creds
+    properties = new HashMap<>(properties);
+    Optional<PolarisEntity> storageInfoEntity =
+        FileIOUtil.findStorageInfoFromHierarchy(resolvedEntityPath);
+    Map<String, String> credentialsMap =
+        storageInfoEntity
+            .map(
+                storageInfo ->
+                    FileIOUtil.refreshCredentials(
+                        realmId,
+                        entityManager,
+                        credentialVendor,
+                        metaStoreSession,
+                        configurationStore,
+                        identifier,
+                        tableLocations,
+                        storageActions,
+                        storageInfo))
+            .orElse(Map.of());
+
+    // Update the FileIO before we write the new metadata file
```

Comment:
What are we updating here?.. and why mention "metadata file"?

Reply:
I just copied this comment from `refreshIOWithCredentials`, will revise the comment.

Reply:
Done!